firmware/picr22-team-OWO/Core/Src/main.c

```diff
@@ -0,0 +1,879 @@
+				   .servo2 = 6150,
+				   .flat_const = 1,
+				   .int_const = 0.1,
+				   .delimiter = 0};
+
+volatile uint8_t isCommandReceived = 0;
+volatile uint16_t commandless_count = 0;
+
+uint16_t clamp(uint16_t value, uint16_t min, uint16_t max) {
+	if (value > max) {return max;}
+	else if (value < min) {return min;}
+	return value;
+}
+
+
+// Copies the recieved command into the command struct
+void CDC_On_Receive(uint8_t* buffer, uint32_t* length) { // command recieve callback, copies data to command struct
+  if (*length == sizeof(Command)) {
+    memcpy(&command, buffer, sizeof(Command));
+    if (command.delimiter == 0xAAAA) {
+      isCommandReceived = 1;
+      commandless_count = 0;
+    }
+  }
+}
+
+// Initialises the encoders
+void enc_init() {
+	HAL_TIM_Encoder_Start(&htim4, TIM_CHANNEL_1 | TIM_CHANNEL_2); // Motor 1 ?
+	HAL_TIM_Encoder_Start(&htim1, TIM_CHANNEL_1 | TIM_CHANNEL_2); // Motor 2
+	HAL_TIM_Encoder_Start(&htim3, TIM_CHANNEL_1 | TIM_CHANNEL_2); // Motor 3 ?
+}
+
+// Initialises the PWMs
+void pwm_init() {
+	TIM8->CCR2 = 0;
+	TIM8->CCR1 = 0;
+	TIM8->CCR3 = 0;
+
+	TIM2->CCR1 = 4915;
+
+	TIM15->CCR1 = 4875;
+	TIM15->CCR2 = 6150;
+
+    HAL_TIM_PWM_Start(&htim8, TIM_CHANNEL_1);
+    HAL_TIMEx_PWMN_Start(&htim8, TIM_CHANNEL_2);
+    HAL_TIMEx_PWMN_Start(&htim8, TIM_CHANNEL_3);
+
+    HAL_TIM_PWM_Start(&htim2, TIM_CHANNEL_1);
+
+    HAL_TIM_PWM_Start(&htim15, TIM_CHANNEL_1);
+    HAL_TIM_PWM_Start(&htim15, TIM_CHANNEL_2);
+}
+
+// Updates the motor's speed, direction and PI constants from the command struct
+void motor_status_update() {
+	for (uint8_t i = 0; i<3; i++) {
+		if (command.speed[i] >= 0) {
+			motor_status[i].direction = 1;
+			if (command.speed[i] == 1) {command.speed[i] = 2;} // Everybody gangsta till the input speed is 1
+			motor_status[i].target_speed = command.speed[i];
+		} else {
+			motor_status[i].direction = 0;
+			if (command.speed[i] == -1) {command.speed[i] = -2;} // Everybody gangsta till the input speed is -1
+			motor_status[i].target_speed = -(command.speed[i]);
+		}
+		motor_status[i].flat_const = command.flat_const;
+		motor_status[i].int_const = ((float)command.int_const)/10;
+		motor_status[i].der_const = ((float)command.der_const)/10;
+	}
+}
+
+// Converts internal encoder speed to a PWM value
+uint16_t motor_pwm(uint8_t m_id) {
+    // For ease of use
+    Motor_Status m =  motor_status[m_id];
+
+    // Get new position of the encoder
+	uint16_t new_pos = 0;
+    switch (m_id) {
+		case 0:
+			new_pos = TIM3->CNT;
+			break;
+		case 1:
+			new_pos = TIM1->CNT;
+			break;
+		case 2:
+			new_pos = TIM4->CNT;
+			break;
+	}
+    int16_t pos_change = abs((int16_t)new_pos - m.enc_pos);
+
+	// Clear PI-s integral value when the bot is ordered to stop, might cause issues down the line
+	if (motor_status[0].target_speed == 0 && motor_status[1].target_speed == 0 && motor_status[2].target_speed == 0) {
+		m.integral = 0;
+	}
+
+	m.error = m.target_speed - pos_change;
+	m.integral += m.error;
```

Comment:
It's probably useful to clamp integral to avoid integral windup.

https://en.wikipedia.org/wiki/Integral_windup

Reply:
Added a clamp